Make update test script restore database instead of reinserting data

erimatnor opened the pull request:
The update test script compares a TimescaleDB instance, which is
updated to the latest version from an older version, to a cleanly
installed one at the latest version. It does this by inserting the
same data on both and comparing the data and metadata tables. This
comparison breaks if the new version has changed some behavior (e.g.,
chunk indexes might use a different naming scheme), because the
updated version will retain the metadata related to the old behavior
while the latest version will create metadata according to the new
behavior.

To fix this issue, the update script now performs a database backup on
the old version and restores it on the cleanly installed version of
TimescaleDB. This retains the metadata in accordance with the old
behavior, although the new version will create new metadata according
to the new behavior. This has the further benefit of testing the
restore process during upgrades, which is probably the way people will
move to a new version of TimescaleDB if they do not update the
extension with the data in place.

erimatnor commented:
Btw, the refactoring should (hopefully) also make the update test script less flaky.

olofr commented on scripts/test_updates.sh:
```diff
@@ -5,75 +5,134 @@ set -o pipefail
-docker rm -vf timescaledb-updated timescaledb-clean
-
-diff ${PGTEST_TMPDIR}/clean.out ${PGTEST_TMPDIR}/updated.out | tee ${PGTEST_TMPDIR}/update_test.output
+docker_pgdiff timescaledb-updated timescaledb-clean /src/test/sql/updates/test-0.1.1.sql
```
Why is this file named with an old version number ? Should it be renamed ?

erimatnor replied:
You mean `test-0.1.1.sql`? I actually don't know why this file has this weird name as I didn't create it. My guess is that it is reflecting testing a database that has been updated from 0.1.1...

In any case, if we want to rename (which I think make sense), it should probably be a separate PR.

RobAtticus replied:
Yeah it probably should be renamed, we didn't even have a version 0.1.1 so that's strange...

RobAtticus replied:
(A separate PR is fine)

cevian commented on test/sql/updates/test-0.1.1.sql:
```diff
@@ -1,4 +1,8 @@
-\d+ _timescaledb_catalog.*;
+\dt+ _timescaledb_catalog.*;
+\di+ _timescaledb_catalog.*
+-- Do not list sequence details because of potentially different state
```
Can't we just restart the two containers to make the sequence details the same?

cevian replied:
The big problem I see here is that without a `\d+ _timescaledb_catalog.*` then we don't output the catalog tables' columns or indexes. This, I think is critical. If we want to fix this I think we need to `\d+ _timescaledb_catalog.hypertable` for every table in our catalog.

erimatnor replied:
Any suggestion on how to get around this? Listing sequences won't work, because their state will be different in updated and restored.

erimatnor commented:
Updated this PR to add INSERT test after update and restore. However, I had to do a code change in how indexes are created on chunks to make this work. It turns out the order of creating chunk indexes was based on sort on main index OID, rather than schema+name, which resulted in indexes being created in different order on the updated vs restored database. I fixed this to order on schema+name, but that also caused a lot of changes in regression test output, hence the big diff.
